File: rd-framework/src/main/kotlin/com/jetbrains/rider/framework/text/impl/RdTextBuffer.kt
package com.jetbrains.rider.framework.text.impl

import com.jetbrains.rider.framework.*
import com.jetbrains.rider.framework.base.RdDelegateBase
import com.jetbrains.rider.framework.base.RdReactiveBase.Companion.logReceived
import com.jetbrains.rider.framework.impl.RdProperty
import com.jetbrains.rider.framework.text.intrinsics.RdAssertion
import com.jetbrains.rider.framework.text.intrinsics.RdChangeOrigin
import com.jetbrains.rider.framework.text.intrinsics.RdTextBufferChange
import com.jetbrains.rider.framework.text.intrinsics.RdTextBufferState
import com.jetbrains.rider.framework.text.*
import com.jetbrains.rider.util.lifetime.Lifetime
import com.jetbrains.rider.util.reactive.*
import com.jetbrains.rider.util.warn


/**
 * Slave of the text buffer supports a list of changes that were introduced locally and can be rolled back when master buffer reports
 * incompatible change
 */
open class RdTextBuffer(delegate: RdTextBufferState, final override val isMaster: Boolean = true) : RdDelegateBase<RdTextBufferState>(delegate), ITextBuffer {
    private val changesToConfirmOrRollback: MutableList<RdTextChange> = arrayListOf()
    private val textChanged: IOptProperty<RdTextChange> = OptProperty()
    private val _historyChanged: ISignal<RdTextChange> = Signal()
    override val historyChanged: ISource<RdTextChange> get() = _historyChanged

    final override var bufferVersion: TextBufferVersion = TextBufferVersion.INIT_VERSION
        private set

    private val localOrigin: RdChangeOrigin = if (isMaster) RdChangeOrigin.Master else RdChangeOrigin.Slave

    init {
        // disabling mastering, text buffer must resolve conflicts by itself
        (delegatedBy.changes as RdProperty<*>).slave()
    }
    override fun bind(lf: Lifetime, parent: IRdDynamic, name: String) {
        super.bind(lf, parent, name)
        delegatedBy.changes.adviseNotNull(lf) {
            if (it.origin != localOrigin) receiveChange(it)
        }

        // for asserting documents equality
        delegatedBy.assertedSlaveText.compose(delegatedBy.assertedMasterText, { slave, master -> slave to master}).advise(lf) { (s, m) ->
            if (s.masterVersion == m.masterVersion
                && s.slaveVersion == m.slaveVersion
                && s.text != m.text) {
                throw IllegalStateException("Master and Slave texts are different.\nMaster:\n$m\nSlave:\n$s")
            }
        }
    }

    protected open fun receiveChange(textBufferChange: RdTextBufferChange) {
        val (newVersion, remoteOrigin, change) = textBufferChange
        require(remoteOrigin != localOrigin)

        if (change.kind == RdTextChangeKind.Reset) {
            changesToConfirmOrRollback.clear()
        } else {
            if (isMaster) {
                if (newVersion.master != bufferVersion.master) {
                    logReceived.warn { "Rejecting the change '$change'" }
                    // reject the change. we've already sent overriding change
                    return
                }
            } else {
                if (newVersion.slave != bufferVersion.slave) {
                    // rollback the changes and notify external subscribers
                    for (ch in changesToConfirmOrRollback.reversed()) {
                        val reversedChange = ch.reverse()
                        _historyChanged.fire(reversedChange)
                        textChanged.set(reversedChange)
                    }
                } else {
                    // confirm the changes queue
                    changesToConfirmOrRollback.clear()
                }
            }
        }

        // apply
        bufferVersion = newVersion
        _historyChanged.fire(change)
        textChanged.set(change)
    }

    private fun incrementBufferVersion() {
        bufferVersion = if (isMaster) bufferVersion.incrementMaster() else bufferVersion.incrementSlave()
    }

    override fun fire(value: RdTextChange) {
        require(delegatedBy.isBound || bufferVersion == TextBufferVersion.INIT_VERSION)
        if (delegatedBy.isBound) protocol.scheduler.assertThread()

        incrementBufferVersion()
        if (value.kind == RdTextChangeKind.Reset) {
            changesToConfirmOrRollback.clear()
        } else if (!isMaster) {
            changesToConfirmOrRollback.add(value)
        }

        _historyChanged.fire(value)
        sendChange(RdTextBufferChange(bufferVersion, localOrigin, value))
    }

    protected open fun sendChange(value: RdTextBufferChange) {
        delegatedBy.changes.set(value)
    }

    override fun reset(text: String) {
        fire(RdTextChange(RdTextChangeKind.Reset, 0, "", text, text.length))
    }

    override fun advise(lifetime: Lifetime, handler: (RdTextChange) -> Unit) {
        require(delegatedBy.isBound)
        protocol.scheduler.assertThread()

        textChanged.advise(lifetime, handler)
    }

    override fun assertState(allText: String) {
        val assertion = RdAssertion(bufferVersion.master, bufferVersion.slave, allText)
        val assertedTextProp = if (isMaster) delegatedBy.assertedMasterText else delegatedBy.assertedSlaveText
        assertedTextProp.set(assertion)
    }

    override fun toString(): String {
        return "Text Buffer: ($rdid)"
    }
}

File: rd-net/RdFramework/Text/Impl/RdTextBuffer.cs
#if !NET35
using System;
using System.Collections.Generic;
using System.Linq;
using JetBrains.Annotations;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;
using JetBrains.Rd.Text.Impl.Intrinsics;
using JetBrains.Rd.Text.Intrinsics;

namespace JetBrains.Rd.Text.Impl
{
  public class RdTextBuffer : RdDelegateBase<RdTextBufferState>, ITextBufferWithTypingSession
  {
    public bool IsMaster { get; }
    private readonly RdChangeOrigin myLocalOrigin;

    private readonly List<RdTextBufferChange> myChangesToConfirmOrRollback;
    private readonly IViewableProperty<RdTextChange> myTextChanged;

    private TextBufferTypingSession? myActiveSession;

    public bool IsCommitting => myActiveSession != null && myActiveSession.IsCommitting;

    public TextBufferVersion BufferVersion { get; private set; }

    /// <summary>
    /// Slave of the text buffer supports a list of changes that were introduced locally and can be rolled back when master buffer reports incompatible change
    /// </summary>
    public RdTextBuffer() : this(new RdTextBufferState())
    {
    }

    public RdTextBuffer(RdTextBufferState state, bool isMaster = false) : base(state)
    {
      IsMaster = isMaster;
      myTextChanged = new ViewableProperty<RdTextChange>();
      myChangesToConfirmOrRollback = new List<RdTextBufferChange>();
      BufferVersion = TextBufferVersion.InitVersion;
      myLocalOrigin = IsMaster ? RdChangeOrigin.Master : RdChangeOrigin.Slave;

      // disabling mastering, text buffer must resolve conflicts by itself
      ((RdProperty<RdTextBufferChange>) Delegate.Changes).IsMaster = false;
    }

    public override void Bind(Lifetime lf, IRdDynamic parent, string name)
    {
      base.Bind(lf, parent, name);
      Delegate.Changes.AdviseNotNull(lf, change =>
      {
        if (change.Origin == myLocalOrigin) return;
        if (myActiveSession != null && myActiveSession.TryPushRemoteChange(change))
        {
          return;
        }

        ReceiveChange(change);
      });

      Delegate.AssertedMasterText.Compose(lf, Delegate.AssertedSlaveText, Tuple.Create).Advise(lf, tuple =>
      {
        var m = tuple.Item1;
        var s = tuple.Item2;
        if (m.MasterVersion == s.MasterVersion
            && m.SlaveVersion == s.SlaveVersion
            && m.Text != s.Text)
        {
          throw new Exception($"Master and Slave texts are different.\nMaster:\n{m.Text}\nSlave:\n{s.Text}");
        }
      });
    }

    private void ReceiveChange(RdTextBufferChange rdTextBufferChange)
    {
      var newVersion = rdTextBufferChange.Version;
      var change = rdTextBufferChange.Change;
      var side = rdTextBufferChange.Origin;
      if (Mode.IsAssertion) Assertion.Assert(side != myLocalOrigin, "side != mySide");

      var masterVersionRemote = newVersion.Master;
      var slaveVersionRemote = newVersion.Slave;

      if (change.Kind == RdTextChangeKind.Reset)
      {
        ClearState();
      }
      else if (change.Kind == RdTextChangeKind.PromoteVersion)
      {
        if (Mode.IsAssertion) Assertion.Assert(!IsMaster);
        BufferVersion = newVersion;
        return;
      }
      else
      {
        if (IsMaster)
        {
          if (Mode.IsAssertion) Assertion.Assert(myChangesToConfirmOrRollback.Count == 0);
          if (masterVersionRemote != BufferVersion.Master)
          {
            // reject the change. we've already sent overriding change.
            return;
          }
        }
        else
        {
          if (slaveVersionRemote != BufferVersion.Slave)
          {
            // rollback the changes and notify external subscribers
            // don't need to update history here - all reverted changes were already stored before on 'fire' stage
            foreach (var ch in Enumerable.Reverse(myChangesToConfirmOrRollback))
            {
              if (ch.Version.Slave <= slaveVersionRemote)
                break;
              myTextChanged.SetValue(ch.Change.Reverse());
            }
            myChangesToConfirmOrRollback.Clear();
          }
          else
          {
            // confirm the changes queue.
            myChangesToConfirmOrRollback.Clear();
          }
        }
      }

      BufferVersion = newVersion;

      if (!IsMaster || myActiveSession == null || !myActiveSession.IsCommitting)
      {
        myTextChanged.SetValue(change);
      }
    }

    private void ClearState()
    {
      myChangesToConfirmOrRollback.Clear();
    }

    public IScheduler? Scheduler { get; set; }

    public void Fire(RdTextChange change)
    {
      if (Mode.IsAssertion) Assertion.Assert(Delegate.IsBound || BufferVersion == TextBufferVersion.InitVersion);
      if (Delegate.IsBound) Proto.Scheduler.AssertThread();

      if (IsMaster && myActiveSession != null && myActiveSession.IsCommitting)
      {
        return;
      }

      IncrementBufferVersion();
      var bufferChange = new RdTextBufferChange(BufferVersion, myLocalOrigin, change);
      if (change.Kind == RdTextChangeKind.Reset)
      {
        ClearState();
      }
      else if (!IsMaster)
      {
        myChangesToConfirmOrRollback.Add(bufferChange);
      }

      myActiveSession?.TryPushLocalChange(change);

      Delegate.Changes.SetValue(bufferChange);
    }

    private void IncrementBufferVersion()
    {
      BufferVersion = IsMaster ? BufferVersion.IncrementMaster() : BufferVersion.IncrementSlave();
    }

    public void Advise(Lifetime lifetime, Action<RdTextChange> change)
    {
      Assertion.Assert(Delegate.IsBound);
      Proto.Scheduler.AssertThread();
      
      myTextChanged.Advise(lifetime, change);
    }

    public void Reset(string text)
    {
      Fire(new RdTextChange(RdTextChangeKind.Reset, 0, "", text, text.Length));
    }

    public void AssertState(string allText)
    {
      var assertion = new RdAssertion(BufferVersion.Master, BufferVersion.Slave, allText);
      if (IsMaster)
        Delegate.AssertedMasterText.SetValue(assertion);
      else
        Delegate.AssertedSlaveText.SetValue(assertion);
    }

    public ITypingSession StartTypingSession(Lifetime lifetime)
    {
      Assertion.Assert(myActiveSession == null);
      Assertion.Assert(lifetime.IsAlive);

      myActiveSession = new TextBufferTypingSession(this);
      lifetime.OnTermination(() => myActiveSession = null);
      return myActiveSession;
    }

    public class TextBufferTypingSession : ITypingSession
    {
      private enum State
      {
        Opened,
        Committing
      }

      private readonly RdTextBuffer myBuffer;
      private readonly List<RdTextBufferChange> myRemoteChanges = new List<RdTextBufferChange>();
      private readonly List<RdTextChange> myLocalChanges = new List<RdTextChange>();
      private readonly TextBufferVersion myVersionBeforeOpening;
      private readonly TextBufferVersion myInitialBufferVersion;
      private State myState = State.Opened;

      public readonly Signal<RdTextChange> OnLocalChange = new Signal<RdTextChange>();
      public readonly Signal<RdTextChange> OnRemoteChange = new Signal<RdTextChange>();

      public TextBufferTypingSession(RdTextBuffer buffer)
      {
        myBuffer = buffer;
        myInitialBufferVersion = myBuffer.BufferVersion;

        if (buffer.IsMaster)
        {
          myVersionBeforeOpening = myBuffer.BufferVersion;
          myBuffer.Delegate.VersionBeforeTypingSession.Value = myInitialBufferVersion;
        }
        else
        {
          myVersionBeforeOpening = buffer.Delegate.VersionBeforeTypingSession.Value;
        }
      }

      public bool IsCommitting => myState == State.Committing;

      public bool TryPushLocalChange(RdTextChange change)
      {
        if (myState != State.Opened) return false;

        OnLocalChange.Fire(change);
        myLocalChanges.Add(change);
        return true;
      }

      private static int CompareVersions(TextBufferVersion first, TextBufferVersion second)
      {
        if (first.Master != second.Master)
          return first.Master - second.Master;
        return first.Slave - second.Slave;
      }

      public bool TryPushRemoteChange(RdTextBufferChange change)
      {
        if (myState != State.Opened) return false;
        if (!myBuffer.IsMaster && change.Version.Master <= myVersionBeforeOpening.Master) return false;
        if (myBuffer.IsMaster && CompareVersions(change.Version, myVersionBeforeOpening) <= 0) return false;

        OnRemoteChange.Fire(change.Change);
        myRemoteChanges.Add(change);
        return true;
      }

      public void CommitRemoteChanges()
      {
        StartCommitRemoteVersion();
        FinishCommitRemoteVersion();
      }

      public void FinishCommitRemoteVersion()
      {
        if (!myBuffer.IsMaster)
        {
          for (var i = myLocalChanges.Count - 1; i >= 0; i--)
          {
            var change = myLocalChanges[i];
            myBuffer.myTextChanged.SetValue(change.Reverse());
          }
        }

        myBuffer.BufferVersion = myInitialBufferVersion;

        foreach (var bufferChange in myRemoteChanges)
        {
          myBuffer.ReceiveChange(bufferChange);
        }
      }

      public void StartCommitRemoteVersion()
      {
        Assertion.Assert(myState == State.Opened);
        myState = State.Committing;
      }
    }
  }
}
#endif